x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/wired_stream.ts

```diff
@@ -0,0 +1,127 @@
+import { Logger } from '@kbn/logging';
+import { InternalIStorageClient, ApplicationDocument } from '@kbn/storage-adapter';
+import { cloneDeep } from 'lodash';
+import { isResponseError } from '@kbn/es-errors';
+import { State } from './state';
+import { StreamActiveRecord, ValidationResult } from './stream_active_record';
+import { syncWiredStreamDefinitionObjects } from '../helpers/sync';
+import { deleteStreamObjects } from '../stream_crud';
+
+export class WiredStream extends StreamActiveRecord<WiredStreamDefinition> {
+  constructor(definition: WiredStreamDefinition) {
+    super(definition); // Maybe the subclass should double check here that it was really given the right type of definition?
+    // What about the assets?
+  }
+
+  public clone(): StreamActiveRecord<WiredStreamDefinition> {
+    return new WiredStream(cloneDeep(this._definition));
+  }
+
+  protected doUpsert(
+    definition: StreamDefinition,
+    desiredState: State,
+    startingState: State
+  ): void {
+    if (!isWiredStreamDefinition(definition)) {
+      throw new Error('Cannot change stream types');
+    }
+
+    this._definition = definition;
+  }
+
+  protected doDelete(desiredState: State, startingState: State): void {
+    // Maybe remove children for example
+  }
+
+  protected async doValidate(
+    desiredState: State,
+    startingState: State,
+    scopedClusterClient: IScopedClusterClient
+  ): Promise<ValidationResult> {
+    const existsInStartingState = startingState.has(this.definition.name);
+
+    if (!existsInStartingState) {
+      // Check for data stream conflict
+      const dataStreamResult = await scopedClusterClient.asCurrentUser.indices.getDataStream({
+        name: this.definition.name,
+      });
+
+      if (dataStreamResult.data_streams.length !== 0) {
+        return {
+          isValid: false,
+          errors: [
+            `Cannot create wired stream "${this.definition.name}" due to conflict caused by existing data stream`,
+          ],
+        };
+      }
+
+      // Check for index conflict
+      await scopedClusterClient.asCurrentUser.indices
+        .get({
+          index: this.definition.name,
+        })
+        .catch((error) => {
+          if (!(isResponseError(error) && error.statusCode === 404)) {
+            throw error;
+          }
+        });
+
+      return {
+        isValid: false,
+        errors: [
+          `Cannot create wired stream "${this.definition.name}" due to conflict caused by existing index`,
+        ],
+      };
+    }
+
+    return { isValid: true, errors: [] };
+  }
+
+  protected async doCommitUpsert(
+    storageClient: InternalIStorageClient<ApplicationDocument<StreamDefinition>>,
+    logger: Logger,
+    scopedClusterClient: IScopedClusterClient,
+    isServerless: boolean
+  ): Promise<void> {
+    await storageClient.index({
+      id: this.definition.name,
+      document: this.definition,
+    });
+
+    await syncWiredStreamDefinitionObjects({
```

Comment:
I know this is how we do it today, but maybe with this new model we can be more selective in what changed and what is required to be committed so we don't have to update everything wholesale? E.g. right now we always try to update the write index mappings, even if something completely unrelated was changed.

On a more generic level this is about building a state model of the changes to _Elasticsearch objects_, not just _streams objects_. This might also help with the ordering of the updates.

Reply:
See [here](https://github.com/elastic/kibana/pull/211696#discussion_r1975413208), and:
Yes, though I do wonder if we can compute the Elasticsearch object changes based on the diff of the two stream definitions, so that we don't have to model the Elasticsearch objects explicitly 🤔 It would simplify things somewhat

Reply:
yeah, I agree, it might not be necessary to model all of Elasticsearch for this.